### PR TITLE
Preserve query string in language switcher

### LINKED ISSUE #5129: How to preserve query string in language switcher
I have url a url like this `/en/media-center/sector?alias=sector-education-and-scholarships`.

Unfortunately `switch_culture_url` produces `https://localhost:44335/RedirectToLocalizedContent?targetculture=ar-EG&contentItemUrl=%2Fen%2Fmedia-center%2Fsector`, which eliminates the query string values. 

Is there any way to preserve the query string in the redirection?

### PR DESCRIPTION
**Pull request by hishamco**
Fixes #5129

### CODE REVIEW

**Comment by hishamco** [comment]
@jtkech this works too except the querystring get encoded and I got 404 for unknown reason

**Comment by jtkech** [comment]
In `SwitchCultureUrlFilter` you can't do this

            var url = urlHelper.RouteUrl("RedirectToLocalizedContent",
                new
                {
                    area = "OrchardCore.ContentLocalization",
                    targetculture = targetCulture,
                    contentItemUrl = request.Path.Value + request.QueryString.Value,
                });

The mapped route `RedirectToLocalizedContent` doesn't include `contentItemUrl` in its pattern so here `contentItemUrl` will be also passed as a query string parameter with a name and a value on which you can't add the ambient query string. Here we need to add the ambient query parameters as additional route data.

Also for infos the `QueryString` doesn't work as `PathString`, e.g its `ToString()` also call its `ToUriComponent` but that only escape the `#` character. When you build a `QueryString` from a string, the string needs to be already escaped (as in `Request.QueryString`). So, when you append it yourself to a path you have to use a web utility helper that will encode it, or you can use an url helper with query params passed as non escaped route data (as they can be retrieved from `Request.Query`)

I will update soon my PR #5129 to show you what i mean.

**Comment by hishamco** [comment]
I see, FYI I just try it in CCP.cshml

**Comment by sebastienros** [comment]
Isnt' there something like `RawUrl` on the request object?

**Comment by hishamco** [comment]
I'm not sure but I will check

**Comment by jtkech** [comment]
@sebastienros the problem here is that the CCP, to target the `ContentCulturePickerController`, generate an url that also contains query string parameters for `targetCulture` and `contentItemUrl`.

So we need to append to them the ambient query parameters that we want to preserve and that the `ContentCulturePickerController` will need to extract and re-use when doing its redirect.

See #5202

**Comment by hishamco** [comment]
I'm not sure if `ContentCulturePickerController` is needed here, but one question I didn't get a proper answer now, Why the encoded url shows 404?

**Comment by jtkech** [comment]
This is because when the CCP or the `SwitchCultureUrlFilter` filter generates the url, `contentItemUrl` itself is passed as a query parameter on which you append the ambient query string but that will be url encoded including its `?` reserved query delimiter.

Let say we are on this page `https://localhost:44300/about?test=test`

The url generated by the CCP or the `SwitchCultureUrlFilter` liquid filter will be

    /RedirectToLocalizedContent?targetculture=en-US&contentItemUrl=%2Fabout%3Ftest%3Dtest

Then the `RedirectToLocalizedContent` action would see that the`contentItemUrl` query parameter is equal to `/about?test=test`, its decoded value, but the `contentItemUrl` parameter is defined as a `PathString` that is not intended to include a query string. So when it does `LocalRedirect('~' + contentItemUrl)` the pathString is implicitly converted to a string that does an uri encoding including the `?` reserved query string delimiter, so this ends up with a non existing url `/about%3Ftest=test`.

Anyway, even if the query string is not broken, it would never find the localized item because the `contentItemUrl` is not intended to include a query string, it is used to find an item through a collection whose keys never include a query string.

**Comment by hishamco** [comment]
Thanks JT again, I think I will dig into the liquid filter tomorrow morning

**Comment by jtkech** [comment]
Okay cool, look at what i did in #5202 to deal with the query parameters.

**Comment by hishamco** [comment]
> look at what i did in #5202 to deal with the query parameters.

Sure

**Comment by hishamco** [comment]
@jtkech it just works without any extra code, I did my test on CCP

**Comment by jtkech** [comment]
Yes not so easy with all these encoding things ;)

First i think it is better to use `LocalRedirect()` with an escaped url as when we do `LocalRedirect(Url.Ation())` where `Url.Action()` returns an escaped url. That's why we used a `PathString` that call its `ToUriComponent()` when its `ToString()` is executed. But as said, if it contains a query string it doesn't escape it correctly, we need to do it separately.

This because `LocalRedirect()` will call `Url.Content()` [see here](https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs#L56) and then set the `HttpContext.Response.Headers[HeaderNames.Location]` **where it's better to pass escaped uri components**  [see here](https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Infrastructure/LocalRedirectResultExecutor.cs#L63).

That said, looking at the url helper `Content()` code, if there is a `~` it uses the `Request.PathBase` that is a `PathString` on which it adds our url, and then it returns the `.Value` that is the non uri escaped value. Hmm, i think this is wrong here. Our passed url would be ok if already escaped but not the `PathBase` part. They could use the combined `PathString` as is => will be encoded when implicitly converted to a string but again if our url contains a `?` it will be escaped / broken (i think there is no problem to url escape twice unless for some special chars as `?`), maybe this is why they don't do that. But i think they would need to use the encoded `PathBase`, not its `.Value` and then add our already encoded url (after the `~`).

        /// <inheritdoc />
        public virtual string Content(string contentPath)
        {
            if (string.IsNullOrEmpty(contentPath))
            {
                return null;
            }
            else if (contentPath[0] == '~')
            {
                var segment = new PathString(contentPath.Substring(1));
                var applicationPath = ActionContext.HttpContext.Request.PathBase;

                return applicationPath.Add(segment).Value;
            }

            return contentPath;
        }

Sorry for the long story, i will think about this tomorrow, for this maybe better to not use `LocalRedirect()` with a `~` (only used in this controller not elsewhere), maybe we have better to 1st generate the full escaped url by using the `LinkGenerator` and then pass it to `LocalRedirect()` without `~`. I will do some tests tomorrow.

Have a good reading of [this aspnet issue](https://github.com/dotnet/aspnetcore/issues/4919) ;)

---

That said, if you have an item with different localizations, does the `RedirectToLocalizedContent` action triggered by the CCP (after having selected a given culture), redirects you to the right localized item ? I don't think so because the `contentItemUrl` containing query parameters can't be mapped by the `RedirectToLocalizedContent` to an existing localized item . You have to create a localizable item in 2 cultures and then try the CCP or the filter to see if it redirects you to the right localized item.

**Comment by hishamco** [comment]
Thanks for the long story ;)

>  I don't think so because the contentItemUrl containing query parameters can't be mapped by the RedirectToLocalizedContent to an existing localized item

You 're right, the query string in not encoded but I the localization content not retrieved, I will try another attempt by using `LinkGenerator` and see what can I come up with

**Comment by hishamco** [comment]
JT now it works as expected and you could try my branch, I will finalize with little refactoring and cleanup

**Comment by hishamco** [comment]
@dodgy could you please try my branch to confirm if it works, while waiting for JT review

**Comment by jtkech** [comment]
@hishamco very good idea to pass the queryString as a route value, we did this at some point, so simple but i forgot it ;)

Okay, you don't use anymore `LocalRedirect()` with a `~` but you pass a non escaped url, `PathString.Value()` => unescaped component, but okay for the query string because `QueryString.Value` => escaped component, in fact the already escaped value used to build it [see here](https://github.com/dotnet/aspnetcore/blob/master/src/Http/Http.Abstractions/src/QueryString.cs#L39). For infos there is also a `PathString.Add(QueryString)` [see here](https://github.com/dotnet/aspnetcore/blob/d30c407dd0a8e3fb4beeb1ab9c226be24bbdbce4/src/Http/Http.Abstractions/src/PathString.cs#L320) that may be useful.

I need to check it again in details, so, i think i will close my related PR and if you agree work directly on your PR to make suggested changes that you will be able to revert if you don't agree.

Also, doing some tests **with header values with non ascii chars**, this by using a tenant name `Tenanté`, i got another issue about an invalid cookie name `orchauth_Tenanté` in the cookie header. So i will fix first this cookie issue and then work on your PR.

**Comment by hishamco** [comment]
Feel free to commit @jtkech, I will do some unit tests too, BTW I was thinking about `Path.Add()`, but I don't think it works if I need add a prefix such `~` to the current url

**Comment by jtkech** [comment]
Oh yes, and just saw that you still use the tilde `~` char.

**Comment by jtkech** [comment]
Okay thanks, i'm starting to work on your PR by 1st merging last dev.

**Comment by jtkech** [comment]
Oups, sorry you already did the merge

**Comment by hishamco** [comment]
> Okay, you don't use anymore LocalRedirect() with a ~

It's there ;), I just initialize the url with `"~" + contentItemUrl + querystring`

**Comment by hishamco** [comment]
Seems I didn't see your comment before ;)

Just a second I will push another commit and rebase

**Comment by jtkech** [comment]
Oups too late

**Comment by hishamco** [comment]
Seems I'm delay with a fraction of time, coz I have a slow internet connection now :(

**Comment by hishamco** [comment]
I already push my commits it's in the wire ;)

**Comment by hishamco** [comment]
I just abort it before it reach, please hold on I will just add one little commit if you don't mind

**Comment by hishamco** [comment]
It's 1:26 AM here, I will push my commit, and check your changes tomorrow morning

**Comment by jtkech** [comment]
Okay no problem, i will work on it very soon

**Comment by hishamco** [comment]
Or just shoot to me a list of that suggestions that you want and I will do it in case you 're busy with something else

Thanks a lot JT

**Comment by jtkech** [comment]
@hishamco 

- Because we merged dev, if you try it do a full rebuild or a git clean, i needed a rebuild.

- I try it with some `é` char in the tenant prefix, autoroute slug, with or without a trailing slash, with or without a query string.

- In the CCP / filter i pass the `Path.Value` unencoded string that will be bound to the action parameter that is a `PathString` and that will call its `ToUriComponent()` when converted explicitly / implicitly `ToString()`. This prevent an unecessary conversion and maybe double encoding.

- We can't bound to a `QueryString` action parameter, it fails, maybe we can for a `PathString` because it is decorated with a `TypeConverter`. So i pass a string for the queryString. That would be enough because for a `QueryString` the underlying string is already an escaped string, but its `ToString()` that calls its `ToUriComponent()` still escape the `#` char `_value.Replace("#", "%23")` to not be interpreted as a fragment separator. So, i init a `QueryString` inside the controller action.

- I don't use anymore the `~` for redirection, because of the problem we saw, this by just using the `PathBase` ourself. So the final url is the sum of 2 `PathString` plus a `QueryString` resulting in an encoded url string where all uri components are correctly escaped.

      var queryString = new QueryString(queryStringValue);
      var url = HttpContext.Request.PathBase + contentItemUrl + queryString;

    Here i can do that because there are overloaded operators `+` to add 2 `PathString` and be correctly combined, but it returns a `PathString`, but but but there is another operator to add a `QueryString` and that return a string. So the result is a string where all components have called there `ToUriComponent()`. Note: if in the above addtion you remove the `QueryString` the result would be a `PathString`, so you would to use e.g

            var url1 = (HttpContext.Request.PathBase + contentItemUrl).ToString();
            var url2 = HttpContext.Request.PathBase.Add(contentItemUrl).ToString();

- Better to check `!contentItemUrl.HasValue` in place of using `string.IsNullOrEmpty()` on it because it ends up with an additional call to `PathString.ToString()` that calls `ToUriComponent()`. Note: when we do `contentItemUrl = "/"` the result is still a `PathString`.

- I saw another little issue with autoroutes, if we define an autoroute with a trailing slash e.g `about/`, it doesn't work, i fixed it here. Note: It fails if we bluid a `PathString` with a string without a leading slash, but a `PathString` can have a trailing slash, this is when there are combined that there are correctly trimmed resulting in a combined `PathString` that also may have a trailing slash.

- Finally i kept your refactorisation of the action controller, but you changed a little the behavior. If the `RedirectToHomePage` option is checked, it will redirects to the homepage but only if it doesn't find previously a localized item as requested. Maybe you didn't see it because to be redirected to this homepage, the homepage itself needs to be localized in the target culture.

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs [hunk 1]
```diff
@@ -34,16 +34,18 @@ public class ContentCulturePickerController : Controller, IUpdateModel
         }
 
         [HttpGet]
-        public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture, PathString contentItemUrl, QueryString queryString)
+        public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture, PathString contentItemUrl, string queryStringValue)
         {
             targetCulture ??= CultureInfo.InvariantCulture.Name;
 
-            if (String.IsNullOrEmpty(contentItemUrl))
+            if (!contentItemUrl.HasValue)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Nice catch, I forgot to revert the change here

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs [hunk 2]
```diff
@@ -34,16 +34,18 @@ public class ContentCulturePickerController : Controller, IUpdateModel
...
             {
                 contentItemUrl = "/";
             }
 
-            var url = "~" + contentItemUrl.Add(queryString);
+            var queryString = new QueryString(queryStringValue);
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Why we need to create a new instance again and again, while we can pass the query string as `string` or `PathString`?!!

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs [hunk 3]
```diff
@@ -63,17 +65,15 @@ public async Task<IActionResult> RedirectToLocalizedContent(string targetCulture
...
                 {
                     // Try to get the Homepage for the current culture
                     var homeLocalizations = await _culturePickerService.GetLocalizationsFromRouteAsync("/");
-                    url = GetLocalizedContentUrl(homeLocalizations);
+                    TryGetLocalizedContentUrl(homeLocalizations);
                 }
             }
 
-            string GetLocalizedContentUrl(IEnumerable<LocalizationEntry> localizationEntries)
+            bool TryGetLocalizedContentUrl(IEnumerable<LocalizationEntry> localizationEntries)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Totally agree

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs [hunk 4]
```diff
@@ -82,10 +82,11 @@ string GetLocalizedContentUrl(IEnumerable<LocalizationEntry> localizationEntries
                     if (localization != null)
                     {
                         url = Url.Action("Display", "Item", new { Area = "OrchardCore.Contents", contentItemId = localization.ContentItemId }) + queryString;
+                        return true;
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Please add new line before

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Views/ContentCulturePickerContainer.cshtml [hunk 5]
```diff
@@ -8,9 +8,9 @@
                 {
                     <a asp-route="RedirectToLocalizedContent"
                        asp-route-area="OrchardCore.ContentLocalization"
-                       asp-route-targetculture="@culture.Name"
-                       asp-route-contentItemUrl="@Context.Request.Path"
-                       asp-route-queryString="@Context.Request.QueryString"
+                       asp-route-targetCulture="@culture.Name"
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.ContentLocalization/Views/ContentCulturePickerContainer.cshtml]: +1

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutoRouteTransformer.cs [hunk 6]
```diff
@@ -20,7 +20,7 @@ public AutoRouteTransformer(IAutorouteEntries entries, IOptions<AutorouteOptions
 
         public override ValueTask<RouteValueDictionary> TransformAsync(HttpContext httpContext, RouteValueDictionary values)
         {
-            if (_entries.TryGetContentItemId(httpContext.Request.Path.Value.TrimEnd('/'), out var contentItemId))
+            if (_entries.TryGetContentItemId(httpContext.Request.Path.Value, out var contentItemId))
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutoRouteTransformer.cs]: IMHO I don't wanna change anything in other modules, so we can sure we didn't break anything else ;)

**Comment by hishamco** [comment]
> We can't bound to a QueryString action parameter

I mean `PathString`

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs (hunk 2 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Now we path a string from the CCP / filter (the underlying string of the original QueryString). Then yes we could just use this string because, as said in my comments, here the underlying string is escaped (not the case for a PathString). But using a `QueryString` allows to use its implicit `ToUriComponent()` that does an additional escape for the `#` char, this is what they do in the aspnetcore code.

But as said we can't define the queryString action parameter as a `QueryString`, it throws an exception, so here it is a string, that's why i do it explicitly in the code. For the contentItemUrl this is the binder that is able to bluid for us the `PathString`.

So, for the queryString we bluid a `QueryString` only once and for the above reason.

See my last comments in the thread.

File: src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs (hunk 4 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.ContentLocalization/Controllers/ContentCulturePickerController.cs]: Okay

**Comment by jtkech** [comment]
We can bind to a `PathString` but not to a `QueryString` as you did in yor last commit

File: src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutoRouteTransformer.cs (hunk 6 above)
Comment (jtkech) [reply src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutoRouteTransformer.cs]: It's safe, the trim is in the autorouteEntries service itself, and it fixes an issue.

**Comment by jtkech** [comment]
@hishamco as a side note, maybe i was not clear, in a full url there are multiple components, e.g hostString, pathString, queryString, fragmentString (we don't need here the last one that is not send to the server), and that have their own `ToUriComponents()` that don't do the same escapings and then that need to be called separately.

That's why it may be hard for a consumer to encode a full but unencoded url string, depending on the parts it includes that it is not aware, this is the provider that would need to encode each part separately.

 E.g if you build a PathString from a string containing a query part, that's not good, when the PathString will be encoded through its `ToString()` or `ToUriComponent()`, e.g it will break the `?` separator.

**Comment by hishamco** [comment]
LGTM after adding a new line before `return true`

:shipit: if it looks great to everyone

**Comment by jtkech** [comment]
Oups, i already did it but forgot to commit

**Comment by hishamco** [comment]
It always happen ;)

Thanks again

**Comment by hishamco** [comment]
@jtkech could we merge this

**Comment by jtkech** [comment]
@hishamco okay i will re-take a look tomorrow
